service-base/src/main/java/org/eclipse/hono/service/impl/RequestResponseEndpoint.java

```diff
@@ -0,0 +1,147 @@
+/**
+ * Copyright (c) 2017 Bosch Software Innovations GmbH.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Bosch Software Innovations GmbH - initial creation
+ */
+package org.eclipse.hono.service.impl;
+
+import io.vertx.core.Vertx;
+import io.vertx.core.eventbus.MessageConsumer;
+import io.vertx.core.json.JsonObject;
+import io.vertx.proton.ProtonQoS;
+import io.vertx.proton.ProtonReceiver;
+import io.vertx.proton.ProtonSender;
+import org.apache.qpid.proton.amqp.transport.AmqpError;
+import org.apache.qpid.proton.codec.DecodeException;
+import org.apache.qpid.proton.message.Message;
+import org.eclipse.hono.config.ServiceConfigProperties;
+import org.eclipse.hono.service.amqp.BaseEndpoint;
+import org.eclipse.hono.util.MessageHelper;
+import org.eclipse.hono.util.ResourceIdentifier;
+
+import java.util.Objects;
+
+import static io.vertx.proton.ProtonHelper.condition;
+
+/**
+ * An abstract base class for implementing endpoints that implement a request response pattern.
+ * <p>
+ * It is used e.g. in the implementation of the device registration and the credentials API endpoints.
+ */
+public abstract class RequestResponseEndpoint<T extends ServiceConfigProperties> extends BaseEndpoint<T> {
+    public static final int REQUEST_RESPONSE_ENDPOINT_DEFAULT_CREDITS = 20;
+
+    private int prefetchReceiver = REQUEST_RESPONSE_ENDPOINT_DEFAULT_CREDITS;
+
+    /**
+     * Creates an endpoint for a Vertx instance.
+     *
+     * @param vertx The Vertx instance to use.
+     * @throws NullPointerException if vertx is {@code null};
+     */
+    protected RequestResponseEndpoint(final Vertx vertx) {
+        super(Objects.requireNonNull(vertx));
+    }
+
+    /**
+     * Checks whether a given message contains all required properties.
+     *
+     * @param linkTarget The resource path to check the message's properties against for consistency.
+     * @param msg The AMQP 1.0 message to perform the checks on.
+     * @return {@code true} if the message passes all checks.
+     */
+    protected abstract boolean verifyMessage(final ResourceIdentifier linkTarget, final Message msg);
```

Comment:
`org.eclipse.hono.server.ForwardingMessageEndpoint` already contains the abstract `passesFormalVerification` method which seems to serve the same purpose. I'd rather *pull up* that method into `BaseEndpoint` instead of re-defining it here. WDYT?